### DISCUSSION TITLE
仕様の詳細について

### DISCUSSION THREAD

**Discussion by kagomen** [lgtm-factory/lgtm-factory#31] [未分類]
そもそも「仕様」についてよくわかってないですが……
Next.jsで開発する！ライブラリはこれ使う！Vercelにデプロイする！みたいなことですか？
それとも、実装する機能についてもしっかり決める感じですかね？👀

**Comment by kazzyfrog** [comment]
仕様って確かに、すごい抽象的なワードですよね！

今回は大きく、以下の２つに分けて、
ざっくり決められるだけ、決められればと思ってます！
- 要求定義：主にユーザー目線で書かれるドキュメント（開発する製品が満たすべき要求を決める）
  - その製品は、何ができる？目的は？
  - ターゲットは誰？
  - なぜ必要？何が便利？
- 要件定義：その製品は、どうやって作るか？主に開発者目線で書かれるドキュメント（要求を形にする）
  - 仕様技術やインフラ
  - 機能
    - スコープ（V1に必要な最低限の機能、V2以降に実装したいアイデア）
  - （デザインや、画面遷移図などの詳細設計）

参考：[要求定義と要件定義の違いを考える](https://qiita.com/sunstripe2011/items/61df719fb1f6178b2605)

**Reply by kagomen** [reply]
なるほど！わかりやすい……！
ありがとうございます🙏✨

要求定義については既にkazzyfrogさんがまとめてくださってるので、そのまま流用する感じでも良さそうですね！
なので、焦点は要件定義になりそう？ですかね👀

**Reply by kazzyfrog** [reply]
ですね！

現状、要求定義を全員で確認しつつ、
要件定義を、ざっくり決めれるだけ、決められればと思ってます！👍

要求定義は、基本的に後から変更しないのに対して、
要件に関しては、実装時に細かく変更することがあると思うので！
デザインとか、テキストとか、ライブラリ選定とか、実装しながら決めたり、柔軟に変更したり！！

**Reply by kagomen** [reply]
了解です！！

**Reply by kazzyfrog** [reply]
確かに、要求は、現状ここまでで、色々書いたメモで良さそう！

あとでREADMEにまとめる感じで！

**Comment by siso001** [comment]
discordで出た機能のアイデアまとめ

> PCに貼るような、キャッチー・オシャレな、ステッカー風

> ポップな感じの方が、こう、受け皿が広くて良いじゃないかな？とは思いました！

> 自分達とか、ユーザーが開発した、サイトとかプロジェクトのアイコンとかから、プロジェクトオリジナルのスタンプにする

> LGTM画像は、コントリビューターがそれぞれ色んなテーマを作れる、って感じのシステムですかね？

> ベースのデザインを用意して、動的な値（色とか、テキストとか、絵文字とか、、）を受け取り、画像を生成して返却する感じのLGTMサービスです！
> そしてユーザーも、自由にベースのデザイン（ステッカー？）を追加できる、オープンソースのプロジェクトにできれば良さそうかな、と思ってますーー！

**Reply by siso001** [reply]
とりあえず
* ベースのデザイン（テンプレート？）を用意して、動的な値を受け取り、画像を返す機能
* ユーザーがベースのデザインを自由に追加できる機能
が最低限の機能ですかね？最初は最低限の機能から実装していくのが良い気がするので、箇条書きにでもまとめておいた方がいいのかな？

**Reply by kagomen** [reply]
最低限の機能としては、

> ベースのデザイン（テンプレート？）を用意して、動的な値を受け取り、画像を返す機能

これだけでも良さそうな気もしますね……！

**Reply by kazzyfrog** [reply]
@siso001 
discordのまとめ、ありがとうございます👍
めっちゃ助かります！！

---

まだジャストアイデアですが、
自分も、最低限は、こんな感じでいいかなと思ってます：
- ベータ版：ベースのデザインを１つ用意して、動的な値を受け取り、画像を返す
- バージョン１：
    - ベースのデザインを複数用意して、動的な値を受け取り、画像を返す。
    - GitHub上から、誰でもデザインを自由に追加できる

その他メモ：
- 画像は、URLコピーで使用できる
- 保存もできた方がいい？？

**Reply by siso001** [reply]
ですね！ベータ版の考え方わかりやすい！ありがとうございます🙏

**Reply by siso001** [reply]
LGTM画像を使って思ったのですが、画像(画像url)を保存するような機能もあると色んな画像を使い分けたい時に便利かもしれない？
実装するならだいぶ後の話になりますが

**Reply by kagomen** [reply]
保存は是非したいですね！！
ファイル保存？とかのあれこれをよく知らないのですが、保存機能を実装しないと右クリック->画像を保存とかでも保存できないんでしょうか？👀

**Reply by siso001** [reply]
作成した画像の一覧をサイト内で見れるようにする感じです！(プライベートな)
ローカルに画像をダウンロードするのとは別です
LGTMeowのように画像をurl＆Markdown形式の文字列をコピペするだけで画像を送れるようにして、その文字列を一覧表示できるように的な🤨
説明が難しい...

**Reply by kazzyfrog** [reply]
要は「お気に入り機能」みたいな感じですよね！

自分が、お気に入りしたコレクション？が見れる的な！
ログイン実装すれば、、いける。。！

**Reply by siso001** [reply]
それです！

**Reply by kagomen** [reply]
なるほど！！理解しました！！！
たしかにあったら便利・・・！

**Comment by kazzyfrog** [comment]
## 要件
暫定版ですが、ざっくり書いてきます！
必要があれば、コメントを追加したり、編集を加えてください！👍

### ユーザーフロー

メインユースケースは、大きく２つに分られるかも：
1. サイトを訪れて、URLコピペして、そのまま使う（自分で編集せず）
2. サイトを訪れて、自分で編集して、URLコピペして、使う

一般的な、LGTM画像サービスは、１だけ。
でも、編集が面倒くさくて、サクッと使いたい人には、簡単にコピペできた方が良い気がする！

1.
- LGTM画像の一覧（動的な値が入力されたランキング？的な）が表示されている
- LGTM画像をクリックすると、GitHubのマークダウンで使える形式の画像URLが、コピーできる
- →GitHub上で、使える！サイトを訪れて、選んで、１タッチ！めっちゃ簡単！

2.
- LGTM画像の一覧が表示されている
- LGTM画像をクリックし、適宜編集する
- GitHubのマークダウンで使える形式の画像URLが、コピーできる
- →GitHub上で、使える！

### 使用技術・インフラ

- React/ Next.js 
- TypeScript
- Tailwind CSS
- フルスタックNext.jsアプリのデプロイ先は、[未定]


### 機能

β：
- ベースのデザインを１つ用意
- 動的な値を受け取り、画像を返すAPIエンドポイント
- 画像は、GitHubのマークダウンで使える形式の画像URLに変換 → コピーボタン
- 最低限のUI: 
    - a

V１：
- ベースのデザインを複数用意して、動的な値を受け取り、画像を返す。
- GitHub上から、誰でもデザインを自由に追加できる

V2以降のアイデア
- a

---

**確定イシュー**はこんな感じかな。。
ベータまでのMust：
- Next.jsの導入
- ベースのデザインを１つ作成
- 画像を返すAPIエンドポイント実装
- URLコピーボタンの実装（GitHubのマークダウンで使える形式の画像URLに変換）
- 動的な値を入力するフォーム？エディタ？の実装
- サイトのフォントの設定（next/font を使用）
- ヘッダーコンポーネントの作成（header.tsx）
- フッターコンポーネントの作成（footer.tsx）
- ボタンコンポーネントの作成（button.tsx）
- （デプロイ関連）

できたら良い：
- Next.jsの特別ファイルの追加
- サイトのメタデータファイルの追加
- READMEを書く
- ESLint, Prettierの設定
    - prettier-plugin-tailwindcss を追加
- huskyとlint-staged
- デザインシステム？カラーパレットを作成するならそれも

**Reply by kazzyfrog** [reply]
デプロイ先は、無料で、サーバーサイドでのAPIリクエスト数の上限が多いところがいいかも！

**Reply by kazzyfrog** [reply]
ページをどうするか

- 各LGTM画像のベースデザインごとに、ページを動的生成するか
- もしくは、アプリの部分はSPAにするか

**Reply by kazzyfrog** [reply]
その他技術スタック：

Mustではない

- ESLint
- Prettier
- huskyとlint-staged
- tailwind 関連
    - radix-ui or shadcn/uiを使うかどうか

**Comment by kazzyfrog** [comment]
## アイデア雑記

**Reply by kazzyfrog** [reply]
絵文字やら、フォントやらをどうするか。。デザインごとに変えるか。

**Reply by kazzyfrog** [reply]
LGTM画像のベースデザインを増やしていく計画？戦略？的なのも、ざっくりで考えた方がいいかも！
V１で、何種類用意するのか、など

まあとりあえずは、自分たちが使うためのサービスでも良い！

ベースデザインの、カテゴリーみたいなのがあると良さそう！
お題じゃないけど、そういうのがあった方が、作成しやすいかも

**Reply by kazzyfrog** [reply]
vercelに、簡単にセルフホストできるボタンを用意するのもいいかも。
プライベートの開発チームで使用できる感じ
めっちゃ余裕があれば

<img width="212" alt="スクリーンショット 2024-08-11 2 28 06" src="https://github.com/user-attachments/assets/c21104f4-0f73-4914-81a7-fa7e1b13d5e2">

**Reply by kazzyfrog** [reply]
- ベースデザインを作成した人の、ハンドルネームが掲載される。
`@user`によって作成されました
的なこと書きたいかも

- SNSシェアもしやすい動線を作る（できたら）

**Reply by kagomen** [reply]
> V１で、何種類用意するのか

とりあえず、最低ラインの3種類（クール、ポップ、キュート？）で良い気が！

> 絵文字やら、フォントやらをどうするか。。デザインごとに変えるか。

できればデザインごとに変えたいですよね🌟
フォントに関しては、
欧文フォントだったら複数用意しても、そんなにパフォーマンス的に負担にはならなさそう？
日本語フォントはごくニュートラルなものを一種類用意するだけで良さそう、と思いました！

**Comment by kazzyfrog** [comment]
## おまけ
（この辺は余裕がある人向け）

LGTM Factoryが取り組む領域：
- 開発者の慣習・カルチャー（LGTM）
- コードレビュー
- ソフトスキル（ピープルスキル）
    - テキストコミュニケーション
    - リモートで協力する方法・プラクティス
- 開発チームの一体感・カルチャー作り
- カジュアルなアイデンティティ表現（ラインのスタンプ的な）

コードレビューや、テキストコミュニケーションを、さらに楽しいものにして、
開発チームの雰囲気を良くし、会話のきっかけを作る！
一体感を強める！！

**Comment by kagomen** [comment]
要件見て、Lucide Iconsのサイト構成？が参考になりそう！と思いました！
https://lucide.dev/icons/

使いたいアイテムをクリックしたら下から出てきて、ボタン一個でコピー、動的な値を入力できる編集画面が左サイドにある感じ！

<img width="1260" alt="スクリーンショット 2024-08-11 12 35 42" src="https://github.com/user-attachments/assets/d6dc4150-807f-4e8f-955f-93ac59bd372c">

<img width="1258" alt="スクリーンショット 2024-08-11 12 36 07" src="https://github.com/user-attachments/assets/f6010995-91f6-4adb-bea2-f0fc8a255ab2">

**Reply by kagomen** [reply]
なので、

> ページをどうするか
各LGTM画像のベースデザインごとに、ページを動的生成するか
もしくは、アプリの部分はSPAにするか

に関しては、個人的にはSPA推しかも・・・！

ただ、v1のテーマ編集に関しては編集できる内容が多いので、Lucide Iconsの感じでやるよりかは専用の編集画面ページがあると良いなーと思いました！
Lottie Filesみたいなイメージです！

<img width="1267" alt="スクリーンショット 2024-08-11 12 50 11" src="https://github.com/user-attachments/assets/315a25f8-67cf-4d97-98e5-90214461f39f">

**Comment by kagomen** [comment]
## デプロイ先

> 無料で、サーバーサイドでのAPIリクエスト数の上限が多いところ

---

AIの回答:


Next.jsを使うプロジェクトなら、デプロイ先はVercelがおすすめです！

**VercelがNext.jsプロジェクトに最適な理由**

* **Next.jsの開発元**: VercelはNext.jsの開発元なので、連携が非常にスムーズで、最適化された環境で動作します。
* **サーバーサイドレンダリング（SSR）と静的サイト生成（SSG）への対応**: Next.jsの強みであるSSRとSSGを最大限に活かせます。
* **Edge Functions**: APIルートを簡単に作成でき、サーバーレスで動作するため、低遅延でスケーラブルなAPIを構築できます。
* **使いやすさ**: デプロイが非常に簡単で、GitHub連携など、開発者フレンドリーな機能が充実しています。
* **無料プランの充実**: Hobbyプランでは、月間100万リクエストまで無料で利用できます。

**他の選択肢**

もちろん、Vercel以外にもNext.jsをデプロイできるサービスはあります。

* **Netlify**: Vercelと並んで人気のある静的サイトホスティングサービスです。Next.jsにも対応しており、無料プランもあります。
* **Render**: 比較的新しいサービスですが、Next.jsのデプロイにも対応しており、無料プランも充実しています。

**比較**

| 機能 | Vercel | Netlify | Render |
|---|---|---|---|
| Next.jsとの親和性 | 👍 | ○ | ○ |
| サーバーレス関数 | Edge Functions | Functions | Serverless Functions |
| データベース連携 | PlanetScale, Upstashなど | FaunaDB, Hasuraなど | Upstash, ElephantSQLなど |
| 無料プラン | 月100万リクエスト | 月12万5千リクエスト | 月750時間 |

**結論**

Next.jsプロジェクトであれば、Vercelが最もおすすめです。使いやすさ、パフォーマンス、無料プランの充実度など、総合的に優れています。 

ただし、他のサービスもそれぞれ特徴があるので、比較検討してみるのも良いでしょう。

**Reply by kagomen** [reply]
Workersにもデプロイできるけど、どうなんだろう？

https://zenn.dev/link/comments/311821c414f9b6
https://zenn.dev/arafipro/articles/2024-01-03-prisma-supabase-sample

DBとか使うわけじゃないから、わりといけるのかな？

**Reply by kazzyfrog** [reply]
Cloudflare Pagesも、Next.jsのサーバーでの機能に最近？対応した、っぽいんですよね！
https://developers.cloudflare.com/pages/framework-guides/nextjs/

ちょっとこの辺り、あとで比較してみます！

**Reply by kagomen** [reply]
> あとで比較してみます！

ありがとうございます🙏

たしかに以前言っていた気が・・・！
自分も読んでみますー！
